crates/astria-celestia-jsonrpc-client/src/blob.rs

```diff
@@ -0,0 +1,155 @@
+use jsonrpsee::core::Error;
+
+pub use crate::rpc_impl::blob::NAMESPACE_ID_AVAILABLE_LEN;
+use crate::{
+    rpc_impl::blob::BlobClient as _,
+    Client,
+};
+
+/// A wrapper around the raw blob object submitted to the celestia JSON RPC.
+///
+/// See [`rpc_impl::blob::Blob`] for the whole thing.
+#[derive(Debug, Default)]
+pub struct Blob {
+    pub namespace_id: [u8; NAMESPACE_ID_AVAILABLE_LEN],
+    pub data: Vec<u8>,
+}
+
+impl Blob {
+    /// Construct a new blob with a zero namespace ID and
+    /// empty data.
+    ///
+    /// Note that the default namespace ID (all zeros) falls in the
+    /// reserved range and will result in an error from the JSON RPC
+    /// endpoint. See [`Blob::namespace_id`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set the namespace ID to the given value.
+    ///
+    /// At the moment the length of a celestia namespace is 10 bytes long.
```

Comment:
This is one of the places where celestia could improve their communication because from the perspective of a developer targeting their stack this is wrong-in-practice.

Background: in `celestia-app` their concept of `namespace` is 29 bytes long. But the `namespace` is the concatenation of `namespace_version + namespace_id`. `namespace_version` occupies 1 byte in the beginning, and `namespace_id` uses the remaining 28 bytes.

Have a look at how these constants are named here: https://github.com/celestiaorg/celestia-app/blob/965eafb4357376aec31f84f3628f7703c5587f9a/pkg/appconsts/global_consts.go#L22

So where this becomes wrong-in-practice is when you try to send it 29 arbitrary bytes:

1. the first byte is always 0 at the moment (namespace version 0)
2. the next 18 bytes *must be* set to 0 (convention for namespace version 0)
3. that leaves you with 10 bytes to set your namespace ID (minus the reserved namespace, which is all 0 except for the last byte; basically a "mask" `0.0.0.0.0.0.0.0.0.256`).

So in practice you only have 10 bytes for your namespace ID because anything else is rejected by the system.